diff --git a/CMakeLists.txt b/CMakeLists.txt
index 0d15b7a..9da6b3c 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -165,6 +165,9 @@ set_target_properties(cint PROPERTIES
   VERSION ${cint_VERSION}
   SOVERSION ${cint_SOVERSION}
   LIBRARY_OUTPUT_DIRECTORY ${PROJECT_BINARY_DIR})
+target_include_directories(cint
+  INTERFACE ${PROJECT_BINARY_DIR}/include
+  PUBLIC    ${PROJECT_SOURCE_DIR}/include)
 if(QUADMATH_FOUND)
   target_link_libraries(cint quadmath)
 endif()
